LiS/man/man4/pipemod.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: pipemod.4.man,v 1.1.4.5 2008-04-29 08:33:19 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:19 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: pipemod.4.man,v $
.\" Revision 1.1.4.5  2008-04-29 08:33:19  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.4  2006/12/28 10:42:08  brian
.\" - a bit better make check target
.\"
.\" Revision 1.1.4.3  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH PIPEMOD 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B pipemod
\- \fISTREAMS\fR module supporting flush handling in \fISTREAMS\fR \fBfifos\fR and \fBpipes\fR
.\"
.\"
.SH SYNOPSIS
.HP
.BI "int " pfds[2] ;
.HP
.BI "int pipe(" pfds );
.HP
.BI "int ioctl(" pfds[0] ", " \fRI_PUSH\fP ", " \fR\(lqpipemod\(rq\fP );
.\"
.\"
.SH DESCRIPTION
.PP
The
.B pipemod
module has the relatively simple task of reversing the sense of the
.B FLUSH
flag bits in
.B M_FLUSH
messages sent in
.I STREAMS
.I fifos
and
.IR pipes .
This must happen at the
.I midpoint
of a
.I fifo
or
.IR pipe ,
so that
.B FLUSHR
becomes
.BR FLUSHW ,
and
.B FLUSHW
becomes
.BR FLUSHR .
.B pipemod
does this, and has no other function.
.sp
To be used appropriately, then,
.B pipemod
must be the first module pushed onto a pipe end or a fifo, but it is
only necessary on one end of a pipe.
.sp
.B pipemod
is not needed if flush handling need not be supported, or if it its
function is supported by other means.
.\"
.\"
.SH MODULES
.PP
.B pipemod
is a
.I STREAMS
module named
.IR pipemod .
.\"
.\"
.SH "SEE ALSO"
.PP
.BR pipe "(3), "
.BR fifo "(4), "
.BR STREAMS "(4) "
.\"
.\"
.SH HISTORY
.PP
UNIX\(rg\ System V Release 4 (SVR4)
.[
svr4
.]
.\"
.\"
.SH AUTHOR
.PP
John Boyd, protologos LLC.
.[
$LIST$
.]
.T3


NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/pipemod.c
/*****************************************************************************

 @(#) $RCSfile: pipemod.c,v $ $Name:  $($Revision: 1.1.1.4.4.13 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: pipemod.c,v $
 Revision 1.1.1.4.4.13  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 1.1.1.4.4.12  2007/08/15 04:57:58  brian
 - GPLv3 updates

 *****************************************************************************/

#ident "@(#) $RCSfile: pipemod.c,v $ $Name:  $($Revision: 1.1.1.4.4.13 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] = "$RCSfile: pipemod.c,v $ $Name:  $($Revision: 1.1.1.4.4.13 $) $Date: 2008-04-29 08:33:11 $";

/*
 *  pipemod.c - "pipemod" module
 *
 *  pipemod's only function is to facilitate flush handling at the
 *  midpoint of pipes by switching the sense of the flush flags in
 *  M_FLUSH messages.  In this implementation, we also do flush
 *  handling as well (since this is a module, and modules are expected
 *  to do flush handling...).
 *
 *  Copyright (C) 2000  John A. Boyd Jr.  protologos, LLC
 */

#include <sys/LiS/module.h>	/* should be first */

#include <sys/LiS/config.h>

#include <sys/stream.h>
#include <sys/stropts.h>
#ifdef LINUX
#include <linux/errno.h>
#else
#include <sys/errno.h>
#endif
#include <sys/cmn_err.h>
#include <sys/osif.h>

/*
 *  Some configuration sanity checks
 */
#ifndef PIPEMOD__MOD
#error Not configured
#endif

#ifndef PIPEMOD__ID
#define PIPEMOD__ID 0x8888
#endif

#define  MOD_ID   PIPEMOD__ID
#define  MOD_NAME "pipemod"

/*
 *  function prototypes
 */
static int _RP pipemod_open(queue_t *, dev_t *, int, int, cred_t *);
static int _RP pipemod_close(queue_t *, int, cred_t *);
static int _RP pipemod_put(queue_t *, mblk_t *);

static void flush_module(queue_t *, mblk_t *);

/*
 *  module structure
 */
static struct module_info pipemod_minfo = {
	MOD_ID,				/* id */
	MOD_NAME,			/* name */
	0,				/* min packet size accepted */
	INFPSZ,				/* max packet size accepted */
	10240L,				/* highwater mark */
	512L				/* lowwater mark */
};

static struct qinit pipemod_rinit = {
	pipemod_put,			/* put */
	NULL,				/* service */
	pipemod_open,			/* open */
	pipemod_close,			/* close */
	NULL,				/* admin */
	&pipemod_minfo,			/* info */
	NULL				/* stat */
};

static struct qinit pipemod_winit = {
	pipemod_put,			/* put */
	NULL,				/* service */
	NULL,				/* open */
	NULL,				/* close */
	NULL,				/* admin */
	&pipemod_minfo,			/* info */
	NULL				/* stat */
};

struct streamtab pipemod_info = {
	&pipemod_rinit,			/* read queue init */
	&pipemod_winit,			/* write queue init */
	NULL,				/* lower mux read queue init */
	NULL				/* lower mux write queue init */
};

/*
 *  open
 */
static int _RP
pipemod_open(q, devp, flag, sflag, credp)
	queue_t *q;
	dev_t *devp;
	int flag, sflag;
	cred_t *credp;
{
	queue_t *rq = RD(q);
	char *cp = rq->q_ptr;

#ifdef PIPE_DEBUG
	cmn_err(CE_CONT, "%s_open( 0x%p, 0x%x, 0x%x, 0x%x, ... )\n", MOD_NAME, q, *devp, flag,
		sflag);
#endif

	if (rq->q_ptr == NULL)
		MODGET();
	rq->q_ptr = (void *) ++cp;
	WR(q)->q_ptr = rq->q_ptr;

	qprocson(q);

	return 0;		/* success */
}

/*
 *  close
 */
static int _RP
pipemod_close(q, flag, credp)
	queue_t *q;
	int flag;
	cred_t *credp;
{
#ifdef PIPE_DEBUG
	cmn_err(CE_CONT, "%s_close( 0x%p, 0x%x, ... )\n", MOD_NAME, q, flag);
#endif

	RD(q)->q_ptr = WR(q)->q_ptr = NULL;
	MODPUT();
	qprocsoff(q);

	return 0;		/* success */
}

/*
 *  canonical module flush handling - this could be used in any module
 */
static void
flush_module(q, mp)
	queue_t *q;
	mblk_t *mp;
{
	queue_t *rq = RD(q), *wq = WR(q);

	if (*(mp->b_rptr) & FLUSHW) {
		if (*(mp->b_rptr) & FLUSHBAND)
			flushband(wq, FLUSHDATA, *(mp->b_rptr + 1));
		else
			flushq(wq, FLUSHDATA);
	}
	if (*(mp->b_rptr) & FLUSHR) {
		if (*(mp->b_rptr) & FLUSHBAND)
			flushband(rq, FLUSHDATA, *(mp->b_rptr + 1));
		else
			flushq(rq, FLUSHDATA);
	}
	if (!SAMESTR(q)) {
		char flush = *(mp->b_rptr) & FLUSHRW;

		switch (flush & FLUSHRW) {
		case FLUSHR:
		case FLUSHW:
			*(mp->b_rptr) = (flush & ~FLUSHRW) | (flush ^ FLUSHRW);
		default:
			break;
		}
	}
	putnext(q, mp);
}

/*
 *  put
 */
static int _RP
pipemod_put(q, mp)
	queue_t *q;
	mblk_t *mp;
{
#ifdef PIPE_DEBUG
	cmn_err(CE_CONT, "%s_wput( 0x%08x, 0x%08x )\n", MOD_NAME, q, mp);
#endif

	switch (mp->b_datap->db_type) {
	case M_FLUSH:
		flush_module(q, mp);
		break;
	default:
		putnext(q, mp);
		break;
	}
	return (0);
}

#ifdef MODULE

/*
 *  Linux loadable module interface
 */

#ifdef KERNEL_2_5
int
pipemod_init_module(void)
#else
int
init_module(void)
#endif
{
	int ret = lis_register_strmod(&pipemod_info, MOD_NAME);

	if (ret < 0) {
		cmn_err(CE_CONT, "%s - unable to register module.\n", MOD_NAME);
		return ret;
	}

	return 0;
}

#ifdef KERNEL_2_5
void
pipemod_cleanup_module(void)
#else
void
cleanup_module(void)
#endif
{
	if (lis_unregister_strmod(&pipemod_info) < 0)
		cmn_err(CE_CONT, "%s - unable to unregister module.\n", MOD_NAME);
	return;
}

#ifdef KERNEL_2_5
module_init(pipemod_init_module);
module_exit(pipemod_cleanup_module);
#endif

#if defined(LINUX)		/* linux kernel */

#ifdef __attribute_used__
#undef __attribute_used__
#endif
#define __attribute_used__

#if defined(MODULE_LICENSE)
MODULE_LICENSE("GPL");
#endif
#if defined(MODULE_AUTHOR)
MODULE_AUTHOR("John Boyd <jaboydjr@protologos.net>");
#endif
#if defined(MODULE_DESCRIPTION)
MODULE_DESCRIPTION("STREAMS 'pipemod' pipe flushing module");
#endif
#if defined(MODULE_ALIAS)
MODULE_ALIAS("streams-" __stringify(LIS_OBJNAME));
#endif

#endif				/* LINUX */
#endif				/* MODULE */
